File: 0014_longest_common_prefix/longest_common_prefix.py
from typing import List


# class Solution:
#     """Поиск строки в подстроке"""
#     def longestCommonPrefix(self, strs: List[str]) -> str:
#         if len(strs) <= 1:
#             return strs[0]
#         min = strs[0]
#         for word in strs[1:]:
#             if len(min) > len(word):
#                 min = word
#         strs.remove(min)
#         i = len(min)
#         while i != 0:
#             flag = True
#             for j in range(len(min) + 1 - i):
#                 prefix = min[j:j+i]
#                 for word in strs:
#                     if prefix not in word:
#                         flag = False
#                 if flag:
#                     return prefix
#             i -= 1
#         return ''

class Solution:
    """Поиск самого длинного префикса в списке строк"""
    def longestCommonPrefix(self, strs: List[str]) -> str:
        min = strs[0]
        for word in strs[1:]:
            if len(min) > len(word):
                min = word
        strs.remove(min)
        i = len(min)
        while i != 0:
            flag = True
            for word in strs:
                if min[0:i] not in word[0:i]:
                    flag = False
                    break
            if flag:
                return min[0:i]
            i -= 1
        return ''


if __name__ == '__main__':
    # x = ["reflower", "flow", "flight"]
    x = ["flower", "flow", "flight"]
    # x = ["dog", "racecar", "car"]
    result = Solution()
    print(result.longestCommonPrefix(x))  # "fl"

File: 0014_longest_common_prefix/longest_common_prefix.js
/**
 * @param {string[]} strs
 * @return {string}
 */
var longestCommonPrefix = function(strs) {
    let min = strs[0]
    for (let i=1; i<strs.length; i++){
        if (min.length > strs[i].length){
            min = strs[i]
        }
    }
    // i = strs.indexOf(min)
    strs.splice(strs.indexOf(min), 1)
    let i = min.length
    while (i !==0){
        let flag = true
        for (let word=0; word<strs.length; word++){
            if (strs[word][0, i].includes(min[0, i], 0)){
                flag = false
                break
            }
        } if( flag == true){
            return min[0, i]
        }
        i--
    }
};

console.log(longestCommonPrefix(["flower","flow","flight"])) // "fl"
console.log(longestCommonPrefix(["dog","racecar","car"])) // ""